waspc/src/Wasp/Generator/DbGenerator.hs

```diff
@@ -99,7 +110,49 @@ genPrismaSchema wasp = createTemplateFileDraft dstPath tmplSrcPath (Just templat
 
 genMigrationsDir :: Wasp -> Maybe FileDraft
 genMigrationsDir wasp =
-  (getMigrationsDir wasp) >>= \waspMigrationsDir ->
+  getMigrationsDir wasp >>= \waspMigrationsDir ->
     Just $ createCopyDirFileDraft (SP.castDir genProjectMigrationsDir) (SP.castDir waspMigrationsDir)
   where
     genProjectMigrationsDir = dbRootDirInProjectRootDir </> dbMigrationsDirInDbRootDir
+
+writeDbSchemaChecksumToFile :: Path' Abs (Dir ProjectRootDir) -> IO ()
+writeDbSchemaChecksumToFile genProjectRootDir = do
+  dbSchemaExists <- doesFileExist dbSchemaFp
+  when dbSchemaExists $ do
+    checksum <- calculateChecksumStringForFile dbSchemaFp
+    writeFile dbSchemaChecksumFp checksum
+  where
+    dbSchemaFp = SP.fromAbsFile $ genProjectRootDir </> dbSchemaFileInProjectRootDir
+    dbSchemaChecksumFp = SP.fromAbsFile $ genProjectRootDir </> dbSchemaChecksumFileInProjectRootDir
+
+checkIfDbSchemaMatchesChecksum :: Path' Abs (Dir ProjectRootDir) -> ExceptT String IO ()
+checkIfDbSchemaMatchesChecksum genProjectRootDir = do
+  dbSchemaExists <- liftIO $ doesFileExist dbSchemaFp
+  let maybeDbSchemaFp = if dbSchemaExists then Just dbSchemaFp else Nothing
+  dbSchemaChecksumExists <- liftIO $ doesFileExist dbSchemaChecksumFp
+  let maybeDbSchemaChecksumFp = if dbSchemaChecksumExists then Just dbSchemaChecksumFp else Nothing
+  checkIfDbSchemaMatchesChecksum' maybeDbSchemaFp maybeDbSchemaChecksumFp
+  where
+    dbSchemaFp = SP.fromAbsFile $ genProjectRootDir </> dbSchemaFileInProjectRootDir
+    dbSchemaChecksumFp = SP.fromAbsFile $ genProjectRootDir </> dbSchemaChecksumFileInProjectRootDir
+
+checkIfDbSchemaMatchesChecksum' :: Maybe FilePath -> Maybe FilePath -> ExceptT String IO ()
+checkIfDbSchemaMatchesChecksum' (Just dbSchemaFp) (Just dbSchemaChecksumFp) = do
+  dbSchemaFileChecksum <- liftIO $ calculateChecksumStringForFile dbSchemaFp
+  dbChecksumFileContents <- liftIO $ readFile dbSchemaChecksumFp
+  if dbSchemaFileChecksum == dbChecksumFileContents
+    then return ()
+    else throwError "Your Prisma schema has changed, you should run wasp db migrate-dev."
+checkIfDbSchemaMatchesChecksum' (Just _dbSchemaFp) Nothing =
+  -- Cannot throw the error below because they may have no entities in the schema.prisma file.
+  -- throwError "You are missing the Prisma schema checksum. Please run wasp db migrate-dev."
+  return ()
+checkIfDbSchemaMatchesChecksum' Nothing (Just _dbSchemaChecksumFp) =
+  throwError "Your .wasp output directory appears malformed. Please clean by running wasp clean."
+checkIfDbSchemaMatchesChecksum' Nothing Nothing = return ()
+
+calculateChecksumStringForFile :: FilePath -> IO String
+calculateChecksumStringForFile file = do
+  contents <- BS.readFile file
+  let Hex checksum = checksumFromByteString contents
+  return checksum
```

Comment:
I would move this function to Util and make it return Hex, so it is clear what the String it returns is, and then if the user of this function wants to unwrap that Hex and make it a String or smth else, they can, but I wouldn't do in this function as it means we loose some of the information we have earlier than we need to.

```suggestion
calculateChecksumStringForFile :: FilePath -> IO Hex
calculateChecksumStringForFile file = checksumFromByteString <$> BS.readFile file
```